internal/multistorage/folder.go

```diff
@@ -0,0 +1,458 @@
+package multistorage
+
+import (
+	"fmt"
+	"io"
+	"path"
+
+	"github.com/wal-g/wal-g/internal/multistorage/cache"
+	"github.com/wal-g/wal-g/internal/multistorage/policies"
+	"github.com/wal-g/wal-g/pkg/storages/storage"
+)
+
+func UseAllAliveStorages(folder storage.Folder) error {
+	mf, ok := folder.(*multiFolder)
+	if !ok {
+		return nil
+	}
+
+	storages, err := mf.cache.AllAliveStorages()
+	if err != nil {
+		return fmt.Errorf("select all alive storages in multistorage folder: %w", err)
+	}
+	if len(storages) == 0 {
+		return fmt.Errorf("no alive storages")
+	}
+	mf.storages = changeDirectory(mf.path, storages...)
+	return nil
+}
+
+func UseFirstAliveStorage(folder storage.Folder) error {
+	mf, ok := folder.(*multiFolder)
+	if !ok {
+		return nil
+	}
+
+	firstStorage, err := mf.cache.FirstAliveStorage()
+	if err != nil {
+		return fmt.Errorf("select first alive storage in multistorage folder: %w", err)
+	}
+	if firstStorage == nil {
+		return fmt.Errorf("no alive storages")
+	}
+	mf.storages = changeDirectory(mf.path, *firstStorage)
+	return nil
+}
+
+func UseSpecificStorage(name string, folder storage.Folder) error {
+	mf, ok := folder.(*multiFolder)
+	if !ok {
+		return nil
+	}
+
+	specificStorage, err := mf.cache.SpecificStorage(name)
+	if err != nil {
+		return fmt.Errorf("select storage %q in multistorage folder: %w", name, err)
+	}
+	mf.storages = changeDirectory(mf.path, specificStorage)
+	return nil
+}
+
+func SetPolicies(folder storage.Folder, policies policies.Policies) {
+	if mf, ok := folder.(*multiFolder); ok {
+		mf.policies = policies
+	}
+}
+
+func changeDirectory(path string, storages ...cache.NamedFolder) []cache.NamedFolder {
+	for _, s := range storages {
+		s.Folder = s.Folder.GetSubFolder(path)
+	}
+	return storages
+}
+
+type Folder interface {
+	storage.Folder
+
+	ExistsInFirst(objectRelativePath string) (bool, string, error)
+	ExistsInAny(objectRelativePath string) (bool, string, error)
+	ExistsInAll(objectRelativePath string) (bool, string, error)
+
+	ReadObjectFromFirst(objectRelativePath string) (io.ReadCloser, string, error)
+	ReadObjectFoundFirst(objectRelativePath string) (io.ReadCloser, string, error)
+
+	ListFolderInFirst() (objects []storage.Object, subFolders []storage.Folder, err error)
+	ListFolderWhereFoundFirst() (objects []storage.Object, subFolders []storage.Folder, err error)
+	ListFolderAll() (objects []storage.Object, subFolders []storage.Folder, err error)
+
+	PutObjectToFirst(name string, content io.Reader) error
```

Comment:
Instead of duplicating these functions, let's provide the desired behavior by argument. If these functions are meant to be internal, remove them from the interface.

Reply:
I decided to remove the interface at all - it's not directly used now anyway. If someone needs these specific functions to call, they can use the struct instead.